### DISCUSSION TITLE
Not configure email can't invite new user?

### DISCUSSION THREAD

**Discussion by gaecoli** [getredash/redash#7153] [Q&A]
When i use latest redash code with docker build up server, and i click settings -> users -> new user -> i set user name and user email. I find pending user with a Api key, and i use `redash.xxxx.com/invite/<Api key>?org_slug=default`, it show me `Your invite link has expired.` How to solve it?

**Comment by gaecoli** [comment]
I check out code 
<img width="824" alt="image" src="https://github.com/user-attachments/assets/8a7bec84-0cd9-47f3-9341-1ecb1cd44316">

**Reply by justinclift** [reply]
Interesting.  Is it feasible to configure an email, so that the generation part is happy and hopefully it all works?

As a data point, the smtp2go.com place has a free tier.  That's what I first started with ages ago when trying things out with Redash for the first time. :smile:

Note that if you do decide to try out smtp2go, then these are the settings in your `/opt/redash/env` that should work:

```
REDASH_MAIL_SERVER=mail.smtp2go.com
REDASH_MAIL_DEFAULT_SENDER=email-address@used.as.sender.address
REDASH_MAIL_PORT=2525
REDASH_MAIL_USE_TLS=yes
REDASH_MAIL_USERNAME=email_username_that_you_set_up_in_smtp2go
REDASH_MAIL_PASSWORD=password to match the REDASH_MAIL_USERNAME
```

**Reply by zachliu** [reply]
you can't use `<Api key>` as the invitation token, it won't pass the token verification
but i agree that the error `Your invite link has expired.` is indeed misleading, the actual error is 
```
Failed to verify invite token: <Api key>
...
itsdangerous.exc.BadSignature: No b'.' found in value
```

as @justinclift mentioned, it's not difficult to set up an email and get the actual invitation url
there is just one catch: when running locally, the invitation url doesn't have the port `5001`

`http://localhost/invite/<invitation_token>`

you'll need to manually add the port

`http://localhost:5001/invite/<invitation_token>`

**Reply by gaecoli** [reply]
> Interesting. Is it feasible to configure an email, so that the generation part is happy and hopefully it all works?
> 
> As a data point, the smtp2go.com place has a free tier. That's what I first started with ages ago when trying things out with Redash for the first time. 😄
> 
> Note that if you do decide to try out smtp2go, then these are the settings in your `/opt/redash/env` that should work:
> 
> ```
> REDASH_MAIL_SERVER=mail.smtp2go.com
> REDASH_MAIL_DEFAULT_SENDER=email-address@used.as.sender.address
> REDASH_MAIL_PORT=2525
> REDASH_MAIL_USE_TLS=yes
> REDASH_MAIL_USERNAME=email_username_that_you_set_up_in_smtp2go
> REDASH_MAIL_PASSWORD=password to match the REDASH_MAIL_USERNAME
> ```

Sorry, i slove it. Recently I was testing whether our documents can cover some machines in mainland China, so I encountered some problems. People here don't use email very much, so they usually don't configure email. I deleted the configuration about email and it worked. Thank you for your reply! @justinclift @zachliu

**Reply by justinclift** [reply]
> there is just one catch: when running locally, the invitation url doesn't have the port 5001

Ahhh.  I wonder if that changes when `REDASH_HOST` is defined in the .env file?

Something along the lines of this is what I'm thinking:

REDASH_HOST=http://localhost:5001

Haven't tested it yet, as I'm not yet down at the local co-working centre.  Should be there in about an hour though.

**Reply by gaecoli** [reply]
> > there is just one catch: when running locally, the invitation url doesn't have the port 5001
> 
> Ahhh. I wonder if that changes when `REDASH_HOST` is defined in the .env file?
> 
> Something along the lines of this is what I'm thinking:
> 
> REDASH_HOST=http://localhost:5001
> 
> Haven't tested it yet, as I'm not yet down at the local co-working centre. Should be there in about an hour though.

Maybe we should keep it.

**Reply by justinclift** [reply]
We can probably add REDASH_HOST to the auto-generated contents of `.env` without any problems:

https://github.com/getredash/redash/blob/76eeea1f64c5f4e2ee436ee64eb3e983b394a960/Makefile#L39-L40

**Reply by gaecoli** [reply]
> We can probably add REDASH_HOST to the auto-generated contents of `.env` without any problems:
> 
> https://github.com/getredash/redash/blob/76eeea1f64c5f4e2ee436ee64eb3e983b394a960/Makefile#L39-L40

OK! It's a good idea!

**Reply by justinclift** [reply]
Thinking about it more, we should put it in the compose file directly where the other static values already are:

https://github.com/getredash/redash/blob/d6c6e3bb7aafae89d4d7a83d51896d7a7d234028/compose.yaml#L12-L21

I'll throw together a PR with that shortly. :smile:

**Reply by gaecoli** [reply]
> Thinking about it more, we should put it in the compose file directly where the other static values already are:
> 
> https://github.com/getredash/redash/blob/d6c6e3bb7aafae89d4d7a83d51896d7a7d234028/compose.yaml#L12-L21
> 
> I'll throw together a PR with that shortly. 😄

I support u.

**Comment by zachliu** [comment]
to sum up, two PRs are needed :stuck_out_tongue_winking_eye: 

1. update the error message to reflect the actual error when token verification failed
2. when running locally, the invitation url should include the port

**Reply by gaecoli** [reply]
> to sum up, two PRs are needed 😜
> 
> 1. update the error message to reflect the actual error when token verification failed
> 2. when running locally, the invitation url should include the port

The first point is needed, but No.2 don't need, i look up `master` code, the invitation url will use your configuration `REDASH_HOST` url, so we don't need to add it.

**Reply by justinclift** [reply]
@gaecoli Cool, same thought. :smile:

**Comment by justinclift** [comment]
This is a PR that adds `REDASH_HOST` to the compose file in our development repo, which fixes the missing port number problem with emails generated from our development environment: https://github.com/getredash/redash/pull/7157

It doesn't fix anything with lousy showing of actual error messages though. :wink:  That might be more @zachliu's kind of thing to take a look at?

**Reply by justinclift** [reply]
k, the `REDASH_HOST` part has been merged.  The "improvement of error message" part still needs fixing though.

**Reply by gaecoli** [reply]
> k, the `REDASH_HOST` part has been merged. The "improvement of error message" part still needs fixing though.

U can add a PR to fix it.

**Reply by justinclift** [reply]
Probably more a @zachliu thing than me just yet. :smile:

**Reply by zachliu** [reply]
No problemo. I'll take a look today! :saluting_face:
